lib/Dialect/Ibis/Transforms/IbisContainersToHW.cpp

```diff
@@ -0,0 +1,306 @@
+using namespace ibis;
+
+static llvm::StringRef getScopeRefModuleName(Type type) {
+  return type.cast<ScopeRefType>().getScopeRef().getAttr().strref();
+}
+
+namespace {
+
+// Analysis result for generating the port interface of a container + a bit of
+// port op caching.
+struct ContainerPortInfo {
+  std::unique_ptr<hw::ModulePortInfo> hwPorts;
+
+  // A mapping between the port name and the port op within the container.
+  llvm::DenseMap<StringAttr, InputPortOp> opInputs;
+
+  // A mapping between the port name and the port op within the container.
+  llvm::DenseMap<StringAttr, OutputPortOp> opOutputs;
+
+  ContainerPortInfo() = default;
+  ContainerPortInfo(ContainerOp container) {
+    SmallVector<hw::PortInfo, 4> inputs, outputs;
+
+    // Gather in and output port ops.
+    for (auto input : container.getBodyBlock()->getOps<InputPortOp>()) {
+      opInputs[input.getSymNameAttr()] = input;
+
+      hw::PortInfo portInfo;
+      portInfo.name = input.getSymNameAttr();
+      portInfo.type = cast<PortOpInterface>(input.getOperation()).getPortType();
+      portInfo.dir = hw::ModulePort::Direction::Input;
+      inputs.push_back(portInfo);
+    }
+
+    for (auto output : container.getBodyBlock()->getOps<OutputPortOp>()) {
+      opOutputs[output.getSymNameAttr()] = output;
+
+      hw::PortInfo portInfo;
+      portInfo.name = output.getSymNameAttr();
+      portInfo.type =
+          cast<PortOpInterface>(output.getOperation()).getPortType();
+      portInfo.dir = hw::ModulePort::Direction::Output;
+      outputs.push_back(portInfo);
+    }
+
+    // Sort the in- and output port lists by name to ensure deterministic
+    // ordering.
+    auto cmp = [](const hw::PortInfo *lhs, const hw::PortInfo *rhs) {
+      return lhs->name.getValue().compare(rhs->name.getValue());
+    };
+    llvm::array_pod_sort(inputs.begin(), inputs.end(), cmp);
+    llvm::array_pod_sort(outputs.begin(), outputs.end(), cmp);
+    hwPorts = std::make_unique<hw::ModulePortInfo>(inputs, outputs);
+  }
+};
+
+using ContainerPortInfoMap = llvm::DenseMap<StringAttr, ContainerPortInfo>;
+
+struct ContainerOpConversionPattern : public OpConversionPattern<ContainerOp> {
+  ContainerOpConversionPattern(MLIRContext *ctx,
+                               ContainerPortInfoMap &portOrder)
+      : OpConversionPattern<ContainerOp>(ctx), portOrder(portOrder) {}
+
+  LogicalResult
+  matchAndRewrite(ContainerOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    rewriter.setInsertionPoint(op);
+
+    const ContainerPortInfo &cpi = portOrder.at(op.getSymNameAttr());
+    auto hwMod = rewriter.create<hw::HWModuleOp>(
+        op.getLoc(), op.getSymNameAttr(), *cpi.hwPorts);
+
+    hw::OutputOp outputOp =
+        cast<hw::OutputOp>(hwMod.getBodyBlock()->getTerminator());
+
+    // Replace all of the reads of the inputs to use the input block arguments.
+    for (auto [idx, input] : llvm::enumerate(cpi.hwPorts->getInputs())) {
```

Comment:
I seem to recall a ValueMapper class which pretty much exactly this and can be passed into some function somewhere (called insert block, or something like that which implied block merging) to take care of this functionality.

Reply:
If you're referring to https://github.com/llvm/circt/blob/main/include/circt/Support/ValueMapper.h, not sure i see how that does the same as what dis piece of code does. Maybe if the `hw.output` op already had values assigned to it, yes (then there would be a replacement of SSA values which is what `ValueMapper` was written to do), but that is not the case here.

Reply:
No, I seem to recall something in MLIR upstream.

Reply:
I think it was https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/IRMapping.h that I was thinking of.

Reply:
Interesting - definitely has overlaps with the ValueMapper, with ValueMapper having this additional capability of inferring backedges.